src/stores/game.js

```diff
@@ -165,106 +164,49 @@ export const useGameStore = defineStore('game', {
-          this.cardChosenFromScrap = null;
-          this.playerChoosingFromScrap = null;
-        }
-        if (Object.hasOwnProperty.call(newGame.lastEvent, 'discardedCards')) {
-          this.discardedCards = newGame.lastEvent.discardedCards;
-        } else {
-          this.discardedCards = null;
-        }
-      }
+      const updateGameMap = {
+        id: (newGame) => (this.id = newGame.id),
+        turn: (newGame) => (this.turn = newGame.turn),
+        chat: (newGame) => (this.chat = cloneDeep(newGame.chat)),
+        deck: (newGame) => (this.deck = newGame.deck.map((card) => createGameCard(card))),
+        scrap: (newGame) => (this.scrap = newGame.scrap.map((card) => createGameCard(card))),
+        log: (newGame) => (this.log = cloneDeep(newGame.log)),
+        name: (newGame) => (this.name = newGame.name),
+        p0Ready: (newGame) => (this.p0Ready = newGame.p0Ready),
+        p1Ready: (newGame) => (this.p1Ready = newGame.p1Ready),
+        passes: (newGame) => (this.passes = newGame.passes),
+        players: (newGame) => {
+          this.players = newGame.players.map((player) => setPlayers(player, this.myPNum)) ?? null;
+        },
+        spectatingUsers: (newGame) => (this.spectatingUsers = newGame.spectatingUsers),
+        twos: (newGame) => (this.twos = newGame.twos.map((card) => createGameCard(card))),
+        topCard: (newGame) => (this.topCard = createGameCard(newGame.topCard) ?? null),
+        secondCard: (newGame) => (this.secondCard = createGameCard(newGame.secondCard) ?? null),
+        oneOff: (newGame) => (this.oneOff = createGameCard(newGame.oneOff) ?? null),
+        oneOffTarget: (newGame) => (this.oneOffTarget = createGameCard(newGame.oneOffTarget) ?? null),
+        isRanked: (newGame) => (this.isRanked = newGame.isRanked),
+        currentMatch: (newGame) => (this.currentMatch = newGame.currentMatch),
+        lastEvent: (newGame) => {
+          const lastEventMap = {
+            change: (newGame) => (this.lastEventChange = newGame.lastEvent.change),
+            oneOff: (newGame) => (this.lastEventOneOffRank = newGame.lastEvent.oneOff.rank),
+            oneOffTargetType: (newGame) => (this.lastEventTargetType = newGame.lastEvent.oneOffTargetType),
+            discardedCards: (newGame) => (this.lastEventDiscardedCards = newGame.lastEvent.discardedCards),
+            chosenCard: (newGame) => {
+              this.lastEventCardChosen = newGame.lastEvent.chosenCard;
+              this.playerChoosingFromScrap = newGame.lastEvent.pNum === this.myPNum;
+            },
+          };
+          Object.keys(newGame.lastEvent).forEach((lastEventKey) => {
+            return lastEventMap[lastEventKey] ? lastEventMap[lastEventKey](newGame) : null;
+          });
+        },
+      };
+      //loop through all keys sent through socket message and run corresponding function
+      Object.keys(newGame).forEach((key) => {
+        return updateGameMap[key] ? updateGameMap[key](newGame) : null;
+      });
```

Comment:
Okay so I am _finally_ looking more deeply at this and I'm wondering what's the added value of the loops specifically? Like what if we just replaced the conditionals with optional chaining and skipped looping through keys?

```suggestion
        this.id = newGame.id ?? null;
        this.turn = newGame.turn ?? null;
```
etc? It seems like that would be a little bit shorter and simpler too

Reply:
I can see that working. Maybe I didn’t look deep enough initially but it felt like some of them had multiple operations that were encapsulated in the functions. If it’s simple enough your solution makes more sense imo.

Reply:
There is no added value to the loop using the action map. 

Before I switched it to the action map, the added value of the loop was that it automatically set the store values based on their corresponding newGame values. If you had added a new variable to the store, you would only need to make sure the API was sending it inside newGame and you wouldn't have to mess with updateGame. 

 The if/else blocks were the edge cases where the store value wasn't a direct copy of the newGame value. (running player cards through the card constructor).

That being said I'm not actually sure how often this would be an advantage

Reply:
The action map was much more readable